urllib3/connectionpool.py

```diff
@@ -167,7 +172,7 @@ class HTTPConnectionPool(ConnectionPool, RequestMethods):
     def __init__(self, host, port=None, strict=False,
                  timeout=Timeout.DEFAULT_TIMEOUT, maxsize=1, block=False,
                  headers=None, retries=None,
-                 _proxy=None, _proxy_headers=None,
+                 _proxy=None, _proxy_headers=None, cj=None,
```

Comment:
So, in re: the connection pool having the cookie jar as an instance, the decision there will be related to the decision about instance vs. argument `CookieJar`s (see below). If we're just going to use one or the other, then we can move the jar selection logic into `PoolManager.urlopen()` and just have the `CookieJar` as an instance attribute of that object, passing a single jar into `ConnectionPool.urlopen()`.

If, on the other hand, we want to use both instance and argument jars, then we need to be able to access them separately from within `ConnectionPool.urlopen()`, since cookie handling logic requires us to have both the `Request`-like and `Response`-like objects at the same time. In that case, we'd initialize each `ConnectionPool` with the same `CookieJar` using `PoolManager.connection_pool_kw`, and leave the option open to receive another `CookieJar` in the `.request` kwargs. This implementation also allows setting up single ConnectionPools that can handle cookies independently.

Regardless, the most efficient way to do all the reset of the cookie-handling logic will be inside `ConnectionPool.urlopen()`.